### DISCUSSION TITLE
Technology Stack

### DISCUSSION THREAD

**Discussion by timvahlbrock** [Oeffis/Oeffis#3] [General]
We need to discuss which Technologies we want to use for our application.

**Comment by timvahlbrock** [comment]
## Frontend Proposal: Hybrid App
A hybrid web app based on [Ionic](https://ionicframework.com) & [Capacitor](https://capacitorjs.com).
Advantages:
- Ability to develop for iOS, Android, Web and Desktop at the same time
- Ability to use Frameworks like Angular and React, which many of us are familiar with
- Ability to still use native functionality, at least where supported by the platform
- I'm already familiar with setting up CI/CD and testing environments for Hybrid Apps, so we would be off to a fast start

**Comment by timvahlbrock** [comment]
## Code Organisation Proposal: Monorepo

In Monorepos, all the code is located in one repository. Although lacking _separation of concerns_ might be the first thing one would think of when hearing this, monorepos provide potential to actually reduce duplication and project complexity. This is currently done by companies like Google, Microsoft, Meta, etc. and also by many open source projects. Accordingly, there is a variety of tools that can help with organising code in monorepos like Bazel, Gradle, Lerna and Nx. Unfortunately I haven't yet worked with those.

More info on monorepos and their advantages and challenges, including info about feature support of different tools: https://monorepo.tools

**Reply by timvahlbrock** [reply]
Advantages:
- Single Source of Truth, Knowledge is not spread across repositories
- easier code sharing, no implicit dependency on other repos
- Single CI/CD

Disadvantages:
- Extra effort for Monorepo tooling
- Missing knowledge

**Comment by timvahlbrock** [comment]
## Testing Proposal: Gherkin

Cucumber provides the possibility to write Feature Specifications in plain, but structured englisch (or most other languages) using the Gherkin syntax. The meaning of specific phrases is then defined separately from the definition of the test case.

Example from the Cucumber Documentation:

```gherkin
Feature: Is it Friday yet?
  Everybody wants to know when it's Friday

  Scenario: Sunday isn't Friday
    Given today is Sunday
    When I ask whether it's Friday yet
    Then I should be told "Nope"
```

```js
const assert = require('assert');
const { Given, When, Then } = require('@cucumber/cucumber');

function isItFriday(today) {
  // We'll leave the implementation blank for now
}

Given('today is Sunday', function () {
  this.today = 'Sunday';
});

When('I ask whether it\'s Friday yet', function () {
  this.actualAnswer = isItFriday(this.today);
});

Then('I should be told {string}', function (expectedAnswer) {
  assert.strictEqual(this.actualAnswer, expectedAnswer);
});
```

This can be applied to both unit and integration testing. Cucumber is a test framework for itself, that is available for multiple languages, but that nevertheless might not fit all architectures. However, there are plugins that enable usage of Gherkin files with other testing frameworks.

**Reply by timvahlbrock** [reply]
Source of the example, the 10 minute cucumber/gherkin tutorial: https://cucumber.io/docs/guides/10-minute-tutorial/?lang=javascript (Please note the option to select different programming languages at the top)

**Reply by timvahlbrock** [reply]
Advantages:
- Enforces abstraction of "what's tested" from "how its implemented", which should result in more readable test
- Testing framework can easily be substituted, as only "translations" need to be rewritten
- Test can be read and written by anyone, even persons with little knowledge of the testing framework or programming

Disadvantages:
- Might require extra setup/maintenance effort
- Tests don't necessarily need to be read or written by persons without programming knowledge, as at least at the current point of time we don't have such persons on the team

**Reply by WConen** [reply]
Perhaps, process-related suggestions should (at least) be cross-referenced in the ["Work-Process" discussion](https://github.com/Oeffies/Oeffies/discussions/4).

**Comment by Paulomart** [comment]
# Server / Client Contracts with OpenAPI Specs

Our frontend and backend need to communicate. OpenAPI allows to formulate a contract between both components. All routes and schemas are contained in a single json file. The file can be used to document the api and to generate code for clientside and serverside use.

See [documentation example](https://petstore.swagger.io/) and [json example](https://petstore.swagger.io/v2/swagger.json).

**Reply by timvahlbrock** [reply]
As far as I'm concerned the effort required for this is rather low and the automatic generation of types even should save us time.

**Reply by JaJako** [reply]
Sounds very useful - especially having the specification stored at one defined place and used as documentation simultaneously.

**Comment by Paulomart** [comment]
# Programming language expertise

I think its a good idea to know what programming languages we each know. Everyone can just leave a comment with their expertise.

**Reply by Paulomart** [reply]
- Python
- TypeScript
- JavaScript
- Java
- PHP
- C#

(In order of preference)

**Reply by timvahlbrock** [reply]
- TypeScript/JavaScript
- Java
- (Python)

**Reply by JaJako** [reply]
- Java
- C/C++
- JavaScript
- (Python)
- (C#)

(In order of knowledge)

**Reply by LaraBiss** [reply]
- TypeScript
- Java
- (Python)

**Reply by AconexOfficial** [reply]
- Java
- Kotlin
- (C#)
- (C++)

**Reply by Escowei98** [reply]
- Typescript / Javascript
- Python
- Java
- (C#)

**Reply by dkrakau** [reply]
- Java
- JavaScript
- Python
- C / C++

**Comment by Paulomart** [comment]
# Hosting and Deployment

Our application will have an backend. I suggest the following structure:

Changes in this repository:

* Create a dockerfile for running the backend in prd
* (Maybe also use this container for dev)
* Build the dockerfile in an GitHub action on each commit
* Publish the docker image in GitHub packages

Other changes:

* Create a new repository `infrastructure`
* Create a kubernetes deployment in that repository using kustomize
* Use the published image versions from GitHub packages
* To update the app, create a new PR that bumps the version of the docker image in the `infrastructure`
* Create an GitHub action that will apply the kubernetes deployment on commits

**Benefits**:

* Ensures that all versions of the app have a docker image
* Ensures that infrastructure configuration is not coupled to the apps code 
* Allows to up- and downgrade the app independently
* All the benefits of `configuration as code`
* Changes to the infrastructure just require push access to the repositiory

**Reply by timvahlbrock** [reply]
I don't quite get wether the source code will be bundled in the docker container. If yes, I don't get what "Ensures that infrastructure configuration is not coupled to the apps code" refers to. If no, why not? Wouldn't that make updating the app a lot more easier? How would you up and downgrade the app then? Do you think it will be possible to completely separate the infrastructure from the source code? Would this be applicable to a mono repo approach?

**Reply by Paulomart** [reply]
> I don't quite get wether the source code will be bundled in the docker container.

Only the assets required to run the app are included in the docker **image**.

> Ensures that infrastructure configuration is not coupled to the apps code

Infrastructure configuration are files that are required to deploy the app. Aka manifests for kubernetes. When developing the app the files are not important. When deploying the app the source files for the app are not relevant.

> Wouldn't that make updating the app a lot more easier?

Yes thats why we decouple infrastructure and app code.

> How would you up and downgrade the app then? 

Just adjust the referenced docker image tag in the manifests for kubernetes.

> Do you think it will be possible to completely separate the infrastructure from the source code?

See above

> Would this be applicable to a mono repo approach?

Yes its possible to combine the infrastructure code into the apps repo. But that would couple all infrastructure changes to the app. (also see above)
